Makefile

```diff
@@ -47,43 +27,80 @@ GO_BUILD_LDFLAGS          := $(shell $(GO_BUILD_LDFLAGS_CMD) LDFLAG)
 # for https://github.com/uber/cadence/issues/3514
-ifndef PERSISTENCE_TYPE
-override PERSISTENCE_TYPE = cassandra
-endif
-
-ifndef TEST_RUN_COUNT
-override TEST_RUN_COUNT = 1
-endif
-
+PERSISTENCE_TYPE ?= cassandra
+TEST_RUN_COUNT ?= 1
 ifdef TEST_TAG
 override TEST_TAG := -tags $(TEST_TAG)
 endif
 
-define thriftrwrule
-THRIFTRW_GEN_SRC += $(THRIFT_GENDIR)/go/$1/$1.go
-
-$(THRIFT_GENDIR)/go/$1/$1.go:: $2
-	@mkdir -p $(THRIFT_GENDIR)/go
-	thriftrw --plugin=yarpc --pkg-prefix=$(PROJECT_ROOT)/$(THRIFT_GENDIR)/go/ --out=$(THRIFT_GENDIR)/go $2
+export PATH := $(shell pwd)/$(BUILD)/bin:$(PATH)
+THRIFTRW_BIN = $(BUILD)/bin/thriftrw
+COPYRIGHT_BIN = $(BUILD)/bin/copyright
+MOCKGEN_BIN = $(BUILD)/bin/mockgen
+ENUMER_BIN = $(BUILD)/bin/enumer
+GOIMPORTS_BIN = $(BUILD)/bin/goimports
+GOLINT_BIN = $(BUILD)/bin/golint
+
+# downloads and builds a go-gettable tool, versioned by go.mod, and installs
+# it into the build folder, named the same as the last portion of the URL.
+define get_tool
+@echo "building $(1) into $(BUILD)/bin/$(notdir $(1)) ..."
+@go build -mod=readonly -o $(BUILD)/bin/$(notdir $(1)) $(1)
 endef
 
-$(foreach tsrc,$(THRIFTRW_SRCS),$(eval $(call \
-	thriftrwrule,$(basename $(notdir \
-	$(shell echo $(tsrc) | tr A-Z a-z))),$(tsrc))))
-
-# Automatically gather all srcs
-ALL_SRC := $(shell find . -name "*.go" | grep -v -e Godeps -e vendor \
-	-e ".*/\..*" \
-	-e ".*/_.*" \
-	-e ".*/mocks.*")
-
-# filter out the src files for tools
-TOOLS_SRC := $(shell find ./tools -name "*.go")
-TOOLS_SRC += $(TOOLS_CMD_ROOT)
-
+$(THRIFTRW_BIN): go.mod
+	$(call get_tool,go.uber.org/thriftrw)
+	$(call get_tool,go.uber.org/yarpc/encoding/thrift/thriftrw-plugin-yarpc)
+
+$(COPYRIGHT_BIN): cmd/tools/copyright/licensegen.go
+	go build -o $(COPYRIGHT_BIN) ./cmd/tools/copyright/licensegen.go
+
+$(MOCKGEN_BIN): go.mod
+	$(call get_tool,github.com/golang/mock/mockgen)
+
+$(ENUMER_BIN): go.mod
+	$(call get_tool,github.com/dmarkham/enumer)
+
+$(GOIMPORTS_BIN): go.mod
+	$(call get_tool,golang.org/x/tools/cmd/goimports)
+
+$(GOLINT_BIN): go.mod
+	$(call get_tool,golang.org/x/lint/golint)
+
+
+THRIFT_GENDIR=.gen/go
+THRIFT_SRCS := $(shell find idls -name '*.thrift')
+# concrete targets to build / the "sentinel" go files that need to be produced per thrift file.
+# idls/thrift/thing.thrift -> thing.thrift -> thing -> .gen/go/thing/thing.go
+THRIFT_GEN_SRC := $(foreach tsrc,$(basename $(subst idls/thrift/,,$(THRIFT_SRCS))),$(THRIFT_GENDIR)/$(tsrc)/$(tsrc).go)
+
+# how to generate each thrift file.
+# note that each generated file depends on ALL thrift files - this is necessary because they can import each other.
+$(THRIFT_GEN_SRC): $(THRIFT_SRCS) $(THRIFTRW_BIN)
+	@# .gen/go/thing/thing.go -> thing.go -> "thing " -> thing -> idls/thrift/thing.thrift
+	$(THRIFTRW_BIN) --plugin=yarpc --pkg-prefix=$(PROJECT_ROOT)/$(THRIFT_GENDIR) --out=$(THRIFT_GENDIR) --no-recurse idls/thrift/$(strip $(basename $(notdir $@))).thrift
```

Comment:
nit: maybe split one options per line, for easier readability?

Reply:
can do - I'll suppress and add an echo then, it's sorta space-consuming in the default output